sdk/core/Azure.Core/src/Serialization/ObjectSerializer.cs

```diff
@@ -31,6 +32,42 @@ public abstract class ObjectSerializer
         /// <param name="cancellationToken">The <see cref="CancellationToken"/> to use during serialization.</param>
         public abstract ValueTask SerializeAsync(Stream stream, object? value, Type inputType, CancellationToken cancellationToken);
 
+        /// <summary>
+        /// Convert the provided value to it's binary representation and return it as a <see cref="BinaryData"/> instance.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        /// <param name="inputType">The type of the <paramref name="value"/> to convert.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to use during serialization.</param>
+        /// <returns>The object's binary representation as <see cref="BinaryData"/>.</returns>
+        public BinaryData SerializeToBinaryData(object? value, Type? inputType = default, CancellationToken cancellationToken = default) =>
+            SerializeToBinaryDataInternalAsync(value, inputType, false, cancellationToken).EnsureCompleted();
+
+        /// <summary>
+        /// Convert the provided value to it's binary representation and return it as a <see cref="BinaryData"/> instance.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        /// <param name="inputType">The type of the <paramref name="value"/> to convert.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to use during serialization.</param>
+        /// <returns>The object's binary representation as <see cref="BinaryData"/>.</returns>
+        public async ValueTask<BinaryData> SerializeToBinaryDataAsync(object? value, Type? inputType = default, CancellationToken cancellationToken = default) =>
+            await SerializeToBinaryDataInternalAsync(value, inputType, true, cancellationToken).ConfigureAwait(false);
+
+        private async ValueTask<BinaryData> SerializeToBinaryDataInternalAsync(object? value, Type? inputType, bool async, CancellationToken cancellationToken)
+        {
+            using var stream = new MemoryStream();
+            inputType ??= value?.GetType() ?? typeof(object);
+            if (async)
+            {
+                await SerializeAsync(stream, value, inputType, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                Serialize(stream, value, inputType, cancellationToken);
+            }
+            stream.Position = 0;
+            return BinaryData.FromStream(stream);
```

Comment:
Would it be better to just retrieve the buffer from MemoryStream and wrap it with BinaryData? One less copy.

Reply:
Additionally/alternatively to using knowledge of the Stream being MemoryStream here, I suggested (in https://github.com/Azure/azure-sdk-for-net/pull/16248#pullrequestreview-517032991) of having FromStream(Async) take the slight hit of an as check to avoid the CopyTo logic, when possible.

Basically, it eliminates the need for this comment on all caller code, probably improving average perf.